### PR TITLE
First pass - trivial synchronized blocks

### LINKED ISSUE #8643: Revise synchronized blocks/methods in critical code path to avoid virtual threads pinning

### PR DESCRIPTION
**Pull request by tzolov**
- Convert the "trivial" synchronized block into reentrant locks.

 Resolves #8643

### CODE REVIEW

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/MulticastSendingMessageHandler.java [hunk 1]
```diff
@@ -48,6 +51,8 @@ public class MulticastSendingMessageHandler extends UnicastSendingMessageHandler
 
 	private volatile MulticastSocket multicastSocket;
 
+	private final Lock lock = new ReentrantLock();
```
Comment (artembilan) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/MulticastSendingMessageHandler.java:54]: Same here for the property from super `UnicastSendingMessageHandler`

File: spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/core/Mqttv3ClientManager.java [hunk 2]
```diff
@@ -36,13 +39,16 @@
...
  * @author Artem Bilan
+ * @author Christian Tzolov
  *
  * @since 6.0
  */
 public class Mqttv3ClientManager
 		extends AbstractMqttClientManager<IMqttAsyncClient, MqttConnectOptions>
 		implements MqttCallbackExtended {
 
+	private final Lock lock = new ReentrantLock();
```
Comment (artembilan) [inline spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/core/Mqttv3ClientManager.java:50]: There is one already in the super class.
Please, consider to reuse it making `protected`.

File: spring-integration-websocket/src/main/java/org/springframework/integration/websocket/ClientWebSocketContainer.java [hunk 3]
```diff
@@ -46,13 +48,16 @@
...
  * @author Gary Russell
+ * @author Christian Tzolov
  *
  * @since 4.1
  */
 public final class ClientWebSocketContainer extends IntegrationWebSocketContainer implements SmartLifecycle {
 
 	private static final int DEFAULT_CONNECTION_TIMEOUT = 10;
 
+	private final Lock lock = new ReentrantLock();
```
Comment (artembilan) [inline spring-integration-websocket/src/main/java/org/springframework/integration/websocket/ClientWebSocketContainer.java:59]: Move this into super `IntegrationWebSocketContainer` and reuse it in its impls

**Review by artembilan — CHANGES_REQUESTED** [review]
Mostly this is OK, but really too much and it is easy to miss something.
Please, see if you really were right about changing `tryLock()` to `lock()` everywhere you found it.

@garyrussell ,

as we already know this is easier to review with `?w=1`.

@tzolov ,

let's see next time if can propose the change in baby steps, per use-case. Really hard to review.

File: spring-integration-file/src/main/java/org/springframework/integration/file/locking/FileChannelCache.java [hunk 4]
```diff
@@ -76,7 +76,7 @@ public static FileLock tryLockFor(File fileToLock) throws IOException {
 		FileLock lock = null;
 		if (channel != null) {
 			try {
-				lock = channel.tryLock();
+				lock = channel.lock();
```
[the remark this answers is not in the document]
Comment (tzolov) [reply spring-integration-file/src/main/java/org/springframework/integration/file/locking/FileChannelCache.java:79]: Ah this is a mistake. I've changed a tryLock that was created by me. Will revert this.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/CachingClientConnectionFactory.java [hunk 5]
```diff
@@ -39,11 +41,15 @@
...
  *
  * @author Gary Russell
+ * @author Christian Tzolov
+ *
  * @since 2.2
  *
  */
 public class CachingClientConnectionFactory extends AbstractClientConnectionFactory implements DisposableBean {
 
+	private final Lock lock = new ReentrantLock();
```
[the remark this answers is not in the document]
Comment (tzolov) [reply spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/CachingClientConnectionFactory.java:51]: Even if it does the original code didn't use it, but used the `this` monitor instead. E.g. `lifecycleMonitor` and `this` are 2 different monitors. 
With this first pass I deliberately tried not to optimise but reproduce the original logic and use the exact monitors types.

[the remark this answers is not in the document]
Comment (artembilan) [reply spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/CachingClientConnectionFactory.java:51]: Yeah... Makes sense.
Let's leave it for now as you did!

File: spring-integration-core/src/main/java/org/springframework/integration/graph/IntegrationGraphServer.java [hunk 6]
```diff
@@ -169,35 +178,41 @@ private <T extends IntegrationNode> T enhance(T node) {
...
-		}
-		Map<String, Object> descriptor = new HashMap<>();
-		descriptor.put("provider", "spring-integration");
-		descriptor.put("providerVersion", implementationVersion);
-		descriptor.put("providerFormatVersion", GRAPH_VERSION);
-		String name = this.applicationName;
-		if (name == null) {
-			name = this.applicationContext.getEnvironment().getProperty("spring.application.name");
+	private Graph buildGraph() {
+		this.lock.lock();
```
[the remark this answers is not in the document]
Comment (tzolov) [reply spring-integration-core/src/main/java/org/springframework/integration/graph/IntegrationGraphServer.java:182]: Technically it should work as it is given the preemtive nature of the locks. 
But looking at the getGraph() it seems like the mutex is redundat there. I belive something like this should be correct:
```java
	public Graph getGraph() {
		if (this.graph == null) {
			buildGraph(); // already under mutex
		}
		return this.graph;
	}
```
what do you think?

[the remark this answers is not in the document]
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/graph/IntegrationGraphServer.java:182]: If several threads call this `getGraph()` and encounter with `this.graph == null`, they step into that `buildGraph()` and wait for a `lock`.
When first one is done with building, it releases lock and the next one takes it and start building a fresh graph.
We do that unconditionally in the `rebuild()`, but `getGraph()` should build only once.

Well, let's leave it as you have in the PR right no and revise it later in the separate change!

File: spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java [hunk 7]
```diff
@@ -179,46 +180,58 @@ protected void doStart() {
...
-		var clientManager = getClientManager();
-		if (clientManager == null) {
-			Assert.state(getUrl() != null || connectionOptions.getServerURIs() != null,
-					"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
-			this.client = this.clientFactory.getAsyncClientInstance(getUrl(), getClientId());
-			this.client.setCallback(this);
-			this.client.connect(connectionOptions).waitForCompletion(getCompletionTimeout());
-			this.client.setManualAcks(isManualAcks());
+	private void connect() throws MqttException {
+		this.lock.lock();
```
Comment (artembilan) [inline spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java:184]: Do you really think that connection monitor and topics monitor should be same instance?
I believe there was a reason to have them separately.

**Review by artembilan — CHANGES_REQUESTED** [review]
Just one concern so far I've noticed.

**Review by artembilan — CHANGES_REQUESTED** [review]
Looks like the change in STOMP module has affected somehow tests:
```
StompInboundChannelAdapterWebSocketIntegrationTests > testWebSocketStompClient() FAILED
    java.lang.AssertionError at StompInboundChannelAdapterWebSocketIntegrationTests.java:162
```

File: spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java (hunk 7 above)
Comment (tzolov) [reply spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java:188]: Good point.  
Looked like the parent and child locks always guard the entire method calls, e.g. like `this` monitor.
But after our remark I found one exception ;) So I re-did the conversion for the mqtt classes from scratch.

**Comment by tzolov** [comment]
>StompInboundChannelAdapterWebSocketIntegrationTests > testWebSocketStompClient() FAILED
    java.lang.AssertionError at StompInboundChannelAdapterWebSocketIntegrationTests.java:162
 
It is a bit strange as we have not changed much on the Stomp Inbound channel code. 
Anyway the core of the issue is even stranger as it complains that 

```
Error creating bean with name 'webSocketEventListener' defined in org.springframework.integration.stomp.inbound.StompInboundChannelAdapterWebSocketIntegrationTests$ServerConfig: 
Unsatisfied dependency expressed through method 'webSocketEventListener' parameter 0: 
No qualifying bean of type 'org.springframework.messaging.support.AbstractSubscribableChannel' available: 
expected single matching bean but found 3: clientInboundChannel,clientOutboundChannel,brokerChannel
```

After qualifying explicitly the taskScheduler :
```java
@Bean
public WebSocketStompClient stompClient(@Qualifier("taskScheduler") TaskScheduler taskScheduler) {
...
``` 
and the `clientOutboundChannel` :

```java
@Bean
public ApplicationListener<SessionSubscribeEvent> webSocketEventListener(@Qualifier("clientOutboundChannel") final AbstractSubscribableChannel clientOutboundChannel) {
...
```

The test passes. But i'm no sure if this fixes or masks the core reason that lead to this.

**Review by artembilan — APPROVED** [review]
This is OK with me.

@garyrussell ,
let us know if you are going to review it, or just let it go and come back to specific concern when it raises!

**Comment by artembilan** [comment]
Then I'll merge this as is and we can revise other concerns in other PRs or so.

**Pull request merged**